fairseq/hub_utils.py

```diff
@@ -196,16 +209,17 @@ def binarize(self, sentence: str) -> torch.LongTensor:
 
     def string(self, tokens: torch.LongTensor) -> str:
         return self.tgt_dict.string(tokens)
-
-    def _build_sample(self, src_tokens: torch.LongTensor):
-        assert torch.is_tensor(src_tokens)
-        dataset = self.task.build_dataset_for_inference([src_tokens], [src_tokens.numel()])
-        sample = dataset.collater([dataset[0]])
-        sample = utils.apply_to_sample(
-            lambda tensor: tensor.to(self.device),
-            sample
-        )
-        return sample
+    
+    def _build_batches(self, tokens: List[List[int]]) -> Iterator[Tuple[torch.Tensor, torch.Tensor, torch.Tensor]]:
+        lengths = torch.LongTensor([t.numel() for t in tokens])
+        itr = self.task.get_batch_iterator(
+            dataset=self.task.build_dataset_for_inference(tokens, lengths),
+            max_tokens=self.args.max_tokens,
+            max_sentences=self.args.max_sentences,
+            max_positions=self.max_positions,
+        ).next_epoch_itr(shuffle=False)
+        for batch in itr:
+            yield (batch["id"], batch["net_input"]["src_tokens"], batch["net_input"]["src_lengths"])
```

Comment:
It seems we can just yield the batch directly, then we don't need to reassemble it in `generate()` above

Reply:
Done :+1: